### PR TITLE
Improve team respawn

### PR DESCRIPTION
**Pull request by ChisatoNishikigi73**
## Description
Improved the squad resurrection system so that a player's entire squad is resurrected at the nearest point after death.

## TODO
- [x] Big world 
- [x] [importance]Scene loading
- [x] Optimize

## Type of changes

- [x] Bug fix
- [ ] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

**Comment by Kawaa-qwq** [comment]
@Birdulon I wonder if there are any suggestions for loading scenarios? After each resurrection, the entity will fall because it is not refreshed

File: src/main/java/emu/grasscutter/game/player/TeamManager.java [hunk 1]
```diff
@@ -616,24 +618,69 @@ public void respawnTeam() {
...
         // Packets
         this.getPlayer().sendPacket(new BasePacket(PacketOpcodes.WorldPlayerReviveRsp));
     }
 
+    public double CalculateDistanceInWorld(Position deadPos, Position pointPos){
+        double x = deadPos.getX() - pointPos.getX();
+        double y = deadPos.getY() - pointPos.getY();
+        double z = deadPos.getZ() - pointPos.getZ();
+        return Math.sqrt(x*x + y*y + z*z);
+    }
```
Comment (ChisatoNishikigi73) [inline src/main/java/emu/grasscutter/game/player/TeamManager.java]: Should this be here?

Comment (GanyusLeftHorn) [inline src/main/java/emu/grasscutter/game/player/TeamManager.java]: This already exists, in `Utils::getDist`.

File: src/main/java/emu/grasscutter/game/player/TeamManager.java [hunk 2]
```diff
@@ -616,24 +618,69 @@ public void respawnTeam() {
...
+            distList.add(CalculateDistanceInWorld(deadPos,entry.getPointData().getTranPos()));  // Calculate the distance from the coordinates of death to each point
+
+            if (Double.parseDouble(distList.get(count).toString()) < CalculateDistanceInWorld(deadPos, reSpawnPos)){
+                reSpawnPos = entry.getPointData().getTranPos();
+            }
+            count++;
+        }while(count < PointList.size());
+
+        return reSpawnPos;
+    }
```
Comment (GanyusLeftHorn) [inline src/main/java/emu/grasscutter/game/player/TeamManager.java]: I haven't tested this, but it seems to me like this whole method could be replaced by the much more concise:

```java
public Position getRespawnPosition() {
    var deathPos = this.getPlayer().getPosition();
    int sceneId = this.getPlayer().getSceneId();

    // Get closest trans point to where the player died.
    var respawnPoint = GameData.getScenePointsPerScene().get(sceneId).stream()
        .map(pointId -> GameData.getScenePointEntryById(sceneId, pointId))
        .filter(point -> point.getPointData().getType().equals("SceneTransPoint"))
        .min(((pos1, pos2) -> Double.compare(Utils.getDist(pos1.getPointData().getTranPos(), deathPos), Utils.getDist(pos2.getPointData().getTranPos(), deathPos))));

    return respawnPoint.get().getPointData().getTranPos();
}
```

That would also solve the problem of non-`SceneTransPoint` points in scenes other than `3`, which are currently not handled by your implementation.

**Review by GanyusLeftHorn — CHANGES_REQUESTED** [review]

File: src/main/java/emu/grasscutter/game/player/TeamManager.java (hunk 2 above)
Comment (ChisatoNishikigi73) [reply src/main/java/emu/grasscutter/game/player/TeamManager.java]: .min(((pos1, pos2) -> Double.compare(Utils.getDist(pos1.getPointData().getTranPos(), deathPos), Utils.getDist(pos2.getPointData().getTranPos(), deathPos))));
-->>
.min((Comparator.comparingDouble(pos -> Utils.getDist(pos.getPointData().getTranPos(), deathPos))));

it's ok?

Comment (ChisatoNishikigi73) [reply src/main/java/emu/grasscutter/game/player/TeamManager.java]: Yes, I completely forgot about it and wrote a method myself

**Review by GanyusLeftHorn — APPROVED** [review]
Looks good to me.

**Comment by GanyusLeftHorn** [comment]
Now that I think about it, maybe we want to make sure to only teleport the player to a trans point they actually have unlocked? If so, simply changing

```java
var respawnPoint = GameData.getScenePointsPerScene().get(sceneId).stream()
```

to

```java
var respawnPoint = this.getPlayer().getUnlockedScenePoints(sceneId).stream()
```

should do the trick.

File: src/main/java/emu/grasscutter/game/player/TeamManager.java [hunk 3]
```diff
@@ -616,24 +619,37 @@ public void respawnTeam() {
...
 
         // Packets
         this.getPlayer().sendPacket(new BasePacket(PacketOpcodes.WorldPlayerReviveRsp));
     }
+    public Position getRespawnPosition() {
+        var deathPos = this.getPlayer().getPosition();
+        int sceneId = this.getPlayer().getSceneId();
+
+        // Get the closest trans point to where the player died.
+        var respawnPoint = GameData.getScenePointsPerScene().get(sceneId).stream()
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/player/TeamManager.java]: Please apply @GanyusLeftHorn's changes. (get player's unlocked waypoints)

**Review by KingRainbow44 — DISMISSED** [review]

File: src/main/java/emu/grasscutter/game/player/TeamManager.java (hunk 2 above)
Comment (GanyusLeftHorn) [reply src/main/java/emu/grasscutter/game/player/TeamManager.java]: > it's ok?

Yes, that should work fine.

**Review by memetrollsXD — APPROVED** [review]

**Comment by ChisatoNishikigi73** [comment]
It's possible that the player hasn't opened any points,
It ensures that the player has no place to respawn.

`try{
            this.getPlayer().sendPacket(new PacketPlayerEnterSceneNotify(this.getPlayer(), EnterType.ENTER_TYPE_SELF, EnterReason.Revival, player.getSceneId(), getRespawnPosition()));
            player.getPosition().set(getRespawnPosition());
        }catch(Exception e){
            this.getPlayer().sendPacket(new PacketPlayerEnterSceneNotify(this.getPlayer(), EnterType.ENTER_TYPE_SELF, EnterReason.Revival, 3, GameConstants.START_POSITION));
            player.getPosition().set(GameConstants.START_POSITION);  // If something goes wrong, the resurrection is here
        }`

**Review by GanyusLeftHorn — APPROVED** [review]

File: src/main/java/emu/grasscutter/game/player/TeamManager.java [hunk 4]
```diff
@@ -3,10 +3,11 @@
 import static emu.grasscutter.config.Configuration.*;
 
 import java.util.*;
-
 import dev.morphia.annotations.Entity;
 import dev.morphia.annotations.Transient;
 import emu.grasscutter.GameConstants;
+import emu.grasscutter.data.GameData;
+import emu.grasscutter.data.binout.ScenePointEntry;
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/game/player/TeamManager.java]:
```suggestion
```

unused import

**Review by Birdulon — APPROVED** [review]
Seems fine for overworld.

Domain wipes seem broken regardless, can't respawn/retry, exit button doesn't exit, manually exiting brings you to overworld with dead team, then relogging brings them all to 100% hp. Something to tackle later.

**Comment by ChisatoNishikigi73** [comment]
> Seems fine for overworld.
> 
> Domain wipes seem broken regardless, can't respawn/retry, exit button doesn't exit, manually exiting brings you to overworld with dead team, then relogging brings them all to 100% hp. Something to tackle later.

Yes, I noticed from the beginning, I am going to pull a new PR to fix all the Dungeon issues, but I don't know if this has happened before because I noticed that there is no code for the 3 buttons.

**Pull request merged**